biolink2/example.rkt

```diff
@@ -54,32 +54,51 @@
                  (mat 'close)
                  (printf "Finished processing ~s rows\n" count)))))
 
-;; materialize the (concept curie property value) relation if not present
-(unless (directory-exists? (db-path "concept"))
-  (printf "buildling relation: ~s; ~s\n"
-          (db-path fnin.nodeprop) (db-path "concept"))
-  (let/files ((in (db-path fnin.nodeprop))) ()
-    (materialize-dsv-stream
-      in header.nodeprop
-      ;; input columns
-      '(curie property value) buffer-size (db-path "concept")
-      ;; attribute names, types, and internal key name for indexing
-      '(curie property value) '(string string string) 'id
-      ;; primary table columns (without any sorted-columns)
-      '(((curie property value) . ())
-        ;; index table columns (without any sorted-columns)
-        ((value property id)    . ())))))
+;; materialize a relation if not present
+(define (materialize-relation name fnin header fields types)
+  (unless (directory-exists? (db-path name))
+    (printf "buildling relation: ~s; ~s\n"
+            (db-path fnin) (db-path name))
+    (let/files ((in (db-path fnin))) ()
+      (materialize-dsv-stream
+        in header
+        ;; input columns
+        fields buffer-size (db-path name)
+        ;; attribute names, types, and internal key name for indexing
+        fields types 'id
+        ;; primary table columns (without any sorted-columns)
+        `((,fields . ())
+          ;; index table columns (without any sorted-columns)
+          (,(append (cdr fields) (list 'id)) . ()))))))
+
+(materialize-relation "concept" fnin.nodeprop header.nodeprop
+                      '(curie property value)
+                      '(string string string))
+
+(materialize-relation "predicate" fnin.edge header.edge
+                      '(:id :start :end)
+                      '(string string string))
+
+(materialize-relation "predicateprop" fnin.edgeprop header.edgeprop
+                      '(:id property value)
+                      '(string string string))
 
 (time (let ()
         ;; baseline
         (define-materialized-relation concept 'disk  (db-path "concept"))
+        (define-materialized-relation predicate 'disk  (db-path "predicate"))
+        (define-materialized-relation predicateprop 'disk  (db-path "predicateprop"))
```

Comment:
I was thinking maybe `edge` and `edge-prop` or `edge-attr`.  It still seems awkward, but I can't think of great names.